Add new EncodeM API

fferrandis opened the pull request:
EncodeM API is intented to be used when we have big part to encode and
want to add the abaility of getting and repairing only small blocks
instead of the whole chunk

For instance EncodeM(data, 4096) will create X K+N association of 4kb of
data. Thanks to that, if we only need a subset of the final data, we can
do that, and we can repair it only with 4kb block buffers

fferrandis committed:
Add new EncodeM API

EncodeM API is intented to be used when we have big part to encode and
want to add the abaility of getting and repairing only small blocks
instead of the whole chunk

For instance EncodeM(data, 4096) will create X K+N association of 4kb of
data. Thanks to that, if we only need a subset of the final data, we can
do that, and we can repair it only with 4kb block buffers

fferrandis committed:
Fix bad length of total payload / change tests cases

fferrandis committed:
Add sub tests cases

fferrandis committed:
Add DecodeMatrix and Repair Matrix

fferrandis committed:
Add testing for decode & decodeslow

fferrandis committed:
Add some tests

fferrandis committed:
Add some comments and make linter happier

fferrandis committed:
Compute Matrix bounds

- add a new function to compute boundaries needed to satisfy a range

fferrandis committed:
Create fragment matrix  rebuilt

slaperche-scality commented on backend.go:
```diff
@@ -15,6 +16,110 @@ uint64_t getOrigDataSize(struct fragment_header_s *header) { return header->meta
 uint32_t getBackendVersion(struct fragment_header_s *header) { return header->meta.backend_version; }
 ec_backend_id_t getBackendID(struct fragment_header_s *header) { return header->meta.backend_id; }
 uint32_t getECVersion(struct fragment_header_s *header) { return header->libec_version; }
+int getHeaderSize() { return sizeof(struct fragment_header_s); }
```
any reason for returning `int` instead of `size_t`?

fferrandis replied:
the sizeof of that struct is mainly used as an int in the liberasurecode (mostly, I guess, because its computation is used  in some substraction with signed integer) I kept that in case this code gonna be reinjected in the main repository to stick to the "coding style"

slaperche-scality replied:
Make sense.
Thanks for the explanation

slaperche-scality commented on backend.go:
```diff
@@ -15,6 +16,110 @@ uint64_t getOrigDataSize(struct fragment_header_s *header) { return header->meta
+                          struct encode_chunk_context *ctx)
+{
+  ctx->instance = liberasurecode_backend_instance_get_by_desc(desc);
+  int i;
```
Why C version are we using here?

If C99 or more  recent, we can declare the counter in the `for` loop directly.

slaperche-scality commented on backend.go:
```diff
@@ -175,6 +282,53 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+	C.encode_chunk_prepare(backend.libecDesc, pData, pDataLen, cChunkSize, &ctx)
+
+	wg.Add(int(ctx.number_of_subgroup))
+	var errCounter uint64
```
minor: Any reason for using 64-bit instead of 32-bit? Perf is different?

Otherwise I would go for 32-bit, it's enough and will avoid bug if we ever want to run on ARM :stuck_out_tongue: 

> On ARM, x86-32, and 32-bit MIPS, it is the caller's responsibility to arrange for 64-bit alignment of 64-bit words accessed atomically. 

[Source](https://golang.org/pkg/sync/atomic/#pkg-note-BUG)

slaperche-scality commented on backend.go:
```diff
@@ -175,6 +282,53 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+							backend.libecDesc, ctx.datas, ctx.codings) }},
+					fmt.Errorf("error encoding chunk (%+v encoding failed)", errCounter)
+	}
+	result := make([][]byte, backend.K+backend.M)
```
I wonder if using `make([][]byte, 0, backend.K+backend.M)` (and then `append` in the loop) makes any difference?

On one hand you avoid the useless default initialization (you'll rewrite each item anyway), on the other hand the default init is `0` and the OS always have some zeroed page nearby, so maybe it's faster that way.

To be benched.

slaperche-scality commented on backend.go:
```diff
@@ -175,6 +282,53 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+					fmt.Errorf("error encoding chunk (%+v encoding failed)", errCounter)
+	}
+	result := make([][]byte, backend.K+backend.M)
+	fragLen := ctx.frags_len
```
Why not `fragLen := int(C.int(fctx.frags_len))` directly?

That would avoid the cast repetition when you're accessing the array.

BTW, why the double cast? `int` doesn't work because the C variable is `unsigned`, so you have to cast to `C.int` first?

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +566,76 @@ func BenchmarkDecode(b *testing.B) {
+
+	buf := make([]byte, 1024*1024)
+	for i := 0; i < len(buf); i++ {
+		buf[i] = byte(65 + i%26)
```
use `'A'` instead of 65

slaperche-scality commented on backend.go:
```diff
@@ -311,45 +376,135 @@ func (backend *Backend) EncodeMatrix(data []byte, chunkSize int) (*EncodeData, e
+				C.ulong(piecesize*backend.K), &outlen)
+			mutex.Lock()
+			defer mutex.Unlock()
+			if C.is_null(p) == C.int(1) {
```
Do we need to call into C for that (it's costly)?

Can't we compare `p` against `nil` directly?

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+			// linearized buffer, according the block number we are decoding
+			var outlen C.uint64_t
+			p := C.decode_fast(C.int(backend.K), cFrags, C.int(len(frags)),
+				C.getStrOffset(data, C.int(blockNr*piecesize*backend.K)),
```
Do we need to call into C for that?

Can't we do something like: `unsafe.Pointer(uintptr(data) + blockNr*piecesize*backend.K)`?

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+
+// decode_fast is used when we have all data fragment. Instead of doing a true decoding, we just
+// reassemble all the fragment linearized in a buffer. This is mainly a copy of liberasurecode
+// fragment_to_string function, excepted that we won't do any addionnal allocation
```
s/excepted/except/

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+// inlen is the array size
+// dest is an already allocated buffer where data will be linearized
+// destlen is the buffer size, and hence, the maximum number of bytes linearized
+// outlen is a pointer containing the number of bytes really linearized in dest (always lower or equal to destlen)
```
I would make the parameter name standing out more, maybe like that:

```c
// `k`: is the number of expected data fragment
// `in`: is an array of all frags
// `inlen`: is the array size
// `dest`: is an already allocated buffer where data will be linearized
// `destlen`: is the buffer size, and hence, the maximum number of bytes linearized
// `outlen`: is a pointer containing the number of bytes really linearized in dest (always lower or equal to destlen)
```

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+// it returns dest if nothing went wrong, else null
+char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uint64_t *outlen)
+{
+        int i;
```
`i` could be declared in the `for` loops

slaperche-scality commented on backend.go:
```diff
@@ -15,6 +16,110 @@ uint64_t getOrigDataSize(struct fragment_header_s *header) { return header->meta
+          }
+          memcpy(ptr, dataoffset, len_to_copy);
+        }
+        dataoffset += ctx->chunk_size;
```
Why is this block indented?

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+  ctx->m = m;
+
+  ctx->datas     = calloc(ctx->k, sizeof(char*));
+  ctx->codings   = calloc(ctx->m, sizeof(char*));
```
Don't we check the return value of `calloc`? (and where do we free this?)

Also, it's possible to write it like this: `ctx->datas = calloc(ctx->k, sizeof *ctx->datas);`

The benefit of this pattern is that, if we ever change the type of `ctx->data`, we don't need to change the `sizeof` expression (so it's less error-prone)

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+  char *m_ref[ctx->m];
+
+  int one_cell_size = sizeof(fragment_header_t) + ctx->chunk_size;
+  int i, ret;
```
Ok, seeing that we're using VLA l.712-173 we must be using C99, so I would declare the `i` in the for loop (to limit its scope)

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+			r := C.encode_chunk(backend.libecDesc, pData, pDataLen, &ctx, C.int(nth))
+			if r < 0 {
+				atomic.AddUint64(&errCounter, 1)
+			}
```
Given the info from [Benoit's link](https://dave.cheney.net/high-performance-go-workshop/gopherchina-2019.html#defer_is_expensive_or_is_it
) + the fact here we are in perf-oriented code, I wonder if it's worth to rewrite as:

```go
					r := C.encode_chunk(backend.libecDesc, pData, pDataLen, &ctx, C.int(nth))
					if r < 0 {
							atomic.AddUint64(&errCounter, 1)
					}
					wg.Done()
```

To be checked with your benchmark (if the benefits are negligible, let's keep the code as-is)

slaperche-scality commented on .github/workflows/go.yml:
```diff
@@ -16,6 +16,21 @@ jobs:
     - name: Check out code into the Go module directory
       uses: actions/checkout@v1
 
+    - name: Install Yaml
```
In the commit message:
- s/abaility/ability/

slaperche-scality replied:
s/yaml/yasm/ :wink:

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+        }
+		memset(frags, 0, sizeof(frags));
+		// we start by iterating on all fragment, and ordering according to the fragment index
+		// in the header all data fragment (fragment whose index is lower than k)
```
Shouldn't be indented

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+                }
+        }
+		// if we don't have enough data fragment, we leave this function and will probably
+		// fallback on a true deocodin function
```
Shouldn't be indented

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+            return NULL;
+        }
+
+		// compute how number of bytes will be linearized
```
Shouldn't be indented

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+            tocopy = destlen;
+        }
+
+		// copy in an ordered way all bytes of fragments in the buffer
```
Shouldn't be indented

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+  unsigned int frags_len; // allocating size of each K+M objects
+  int blocksize;          // k-bounds of data
+  int k;
+  int m;
```
Do we want to use those "variable width" (depends on the underlying arch) integers instead of the fixed width one (like `int32_t` and others)?

I guess we have to if it's what return `liberasure`, but I'm asking just in case :slightly_smiling_face:

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,223 @@ package erasurecode
+    if (ret < 0) {
+      fprintf(stderr, "error encode ret = %d\n", ret);
+      return -1;
+    }
```
Since you have a `return` in the `if`, you can get rid of the `else` and save one indentation level:

```c
  ret = ec->common.ops->encode(ec->desc.backend_desc, k_ref, m_ref, ctx->chunk_size);
  if (ret < 0) {
      fprintf(stderr, "error encode ret = %d\n", ret);
      return -1;
  }
  ret = finalize_fragments_after_encode(ec, ctx->k, ctx->m, ctx->chunk_size, ctx->chunk_size, k_ref, m_ref);
  if (ret < 0) {
    fprintf(stderr, "error encode ret = %d\n", ret);
    return -1;
  }
```

slaperche-scality commented on backend.go:
```diff
@@ -23,9 +230,12 @@ import (
 	"unsafe"
 )
 
+// Version describes the module version
```
`the module version`

I don't think you need to mention `Version` again, as it will be shown in the godoc anyway

slaperche-scality commented on backend.go:
```diff
@@ -36,6 +246,7 @@ func (v Version) String() string {
 	return fmt.Sprintf("%d.%d.%d", v.Major, v.Minor, v.Revision)
 }
 
+// Less compares two versions
```
> Check if `other` is greater than the current version object.

?

slaperche-scality commented on backend.go:
```diff
@@ -47,6 +258,7 @@ func (v Version) Less(other Version) bool {
 	return false
 }
 
+// GetVersion return a structure describing the version of the module
```
Ditto, could be simply `Return a structure describing the version of the module`

Same remarks for the other comments (i.e.: I don't think it's necessary to start the doc with the name of the documented object).

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
 
+// GetHeaderSize returns the size of the header written by liberasurecode before each chunks
+func (backend *Backend) GetHeaderSize() int {
+	r := int(C.getHeaderSize())
```
minor: `return int(C.getHeaderSize())`

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+	result := make([][]byte, backend.K+backend.M)
+	fragLen := ctx.frags_len
+	for i := 0; i < backend.K; i++ {
+		result[i] = (*[1 << 30]byte)(unsafe.Pointer(C.getStrArrayItem(ctx.datas, C.int(i))))[:int(C.int(fragLen)):int(C.int(fragLen))]
```
I would abstract this "hack" into an helper function:

```go
func makeCBackedSlice(cArray unsafe.Pointer, size int) []byte {
	// See https://github.com/golang/go/wiki/cgo#turning-c-arrays-into-go-slices
	return (*[1 << 30]byte)cArray[:size:size]
}
```
and then use it here:

```go
ptr := unsafe.Pointer(C.getStrArrayItem(ctx.datas, C.int(i)))
result[i] = makeCBackedSlice(ptr, fragLen)
```

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+	for index, frag := range frags {
+		C.setStrArrayItem(cFrags, C.int(index), (*C.uchar)(&frag[0]))
+	}
+	lenFrags1 := len(frags[0])
```
`fragLen`? (better than the `1` suffix, and consistent with the naming used in `EncodeMatrix`)

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+			p := C.decode_fast(C.int(backend.K), cFrags, C.int(len(frags)),
+				C.getStrOffset(data, C.int(blockNr*piecesize*backend.K)),
+				C.ulong(piecesize*backend.K), &outlen)
+			mutex.Lock()
```
Could use atomic for `errorNb` and `totLen` instead of a mutex?

slaperche-scality commented on backend.go:
```diff
@@ -175,11 +397,184 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
+
+// decodeMatrixSlow is a fallback when something went wrong with decodeMatrix (especially when data part is missing)
+func (backend *Backend) decodeMatrixSlow(frags [][]byte, piecesize int) (*DecodeData, error) {
+	lenData := len(frags[0])
```
`fragLen`? (consistent with the naming used elsewhere)

slaperche-scality commented on backend.go:
```diff
@@ -231,11 +627,45 @@ func (backend *Backend) Reconstruct(frags [][]byte, fragIndex int) ([]byte, erro
+		return nil, errors.New("reconstruction requires at least one fragment")
+	}
+
+	lenData := len(frags[0])
```
ditto, `fragLen`?

slaperche-scality commented on backend_test.go:
```diff
@@ -533,8 +534,75 @@ func BenchmarkEncode(b *testing.B) {
+		if err != nil {
+			b.Fatal(err)
+		}
+		defer encoded.Free()
```
Hum, this will accumulate memory, no?

`defer` are called when the function returns, not at the end of each iteration.
if `N` is big it can be an issue/influence the benchmark.

If the goal was to postpone deallocation out of the benchmark, maybe we can simply surround the calls to the `Free` functions between `b.StopTimer()` and `b.StartTimer()`

Same remarks below.

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+	testParams := [][]int{{4096, 4097}, {4097, 4096}, {4096, len(buf)}, {4096, 4096}}
+
+	for _, param := range testParams {
+		p := param
```
Maybe directly write `for _, p  := range testParams {` then :stuck_out_tongue:

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+
+	// All our sub tests case. Each {X,Y} represents respecitvely the chunking unit (size of each subpart)
+	// and the data we want to read when we decode
+	testParams := [][]int{{0, 95, 32, 0, 112}, {10, 97, 32, 0, 224}, {97, 98, 32, 112, 224}, {0, 32*backend.K*2 + 1 + 1, 32, 0, 336}}
```
Line is a bit long, could be wrapped:

```go
	testParams := [][]int{
		{0, 95, 32, 0, 112},
		{10, 97, 32, 0, 224},
		{97, 98, 32, 112, 224},
		{0, 32*backend.K*2 + 1 + 1, 32, 0, 336},
	}
```

Could also use the `struct` to improve the readability

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+		t.Fatalf("cannot init backend: (%v)", err)
+	}
+
+	// All our sub tests case. Each {X,Y} represents respecitvely the chunking unit (size of each subpart)
```
`Each {X,Y} ` => we have 5 values per test here, not 2.
Outdated comment?

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+
+	// All our sub tests case. Each {X,Y} represents respecitvely the chunking unit (size of each subpart)
+	// and the data we want to read when we decode
+	testParams := [][]int{{4096, 4097}, {4097, 4096}, {4096, len(buf)}, {4096, 4096}}
```
Could use the `struct` trick to make the code more readable (then you don't need the comment, and you can use named field instead of `0` and `1`).

```go
	testParams := []struct {
		chunkingUnit int
		data         int
	}{
		{chunkingUnit: 4096, data: 4097},
		{chunkingUnit: 4097, data: 4096},
		{chunkingUnit: 4096, data: len(buf)},
		{chunkingUnit: 4096, data: 4096},
	}
```

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+
+	buf := make([]byte, 1024*1024)
+	for i := 0; i < len(buf); i++ {
+		buf[i] = byte(65 + i%26)
```
use `'A'` instead of 65

slaperche-scality commented on backend_test.go:
```diff
@@ -550,3 +618,157 @@ func BenchmarkDecode(b *testing.B) {
+	testParams := [][]int{{4096, 0}, {4096, backend.K}, {DefaultChunkSize, 1}}
+
+	for _, param := range testParams {
+		p := param
```
Same remarks as previous tests on this part.

fferrandis committed:
Coding style & use atomic instead of mutexes

fferrandis committed:
remove some CGO call

fferrandis committed:
Avoid call to C.getHeaderSize during data processing

fferrandis committed:
Remove costly C call to getStrArrayItem

fferrandis committed:
Remove useless cFrags allocation and remove some defer

fferrandis committed:
Avoid as musch as possible call to defer

fferrandis committed:
Clean test & make valgrind happier

fferrandis committed:
Replace C setStrArray by a Go implementation

slaperche-scality commented on backend.go:
```diff
@@ -43,9 +42,9 @@ char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uin
-		// in the header all data fragment (fragment whose index is lower than k)
+	memset(frags, 0, sizeof(frags));
+	// we start by iterating on all fragment, and ordering according to the fragment index
+	// in the header all data fragment (fragment whose index is lower than k)
```
Those lines are indented with tab, but the rest of the C code use spaces, which can give a weird layout  depending on the editor configuration.

slaperche-scality commented on backend.go:
```diff
@@ -70,13 +69,13 @@ char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uin
-		// if we don't have enough data fragment, we leave this function and will probably
-		// fallback on a true deocodin function
+	// if we don't have enough data fragment, we leave this function and will probably
+	// fallback on a true deocodin function
```
ditto, tab are used here

slaperche-scality commented on backend.go:
```diff
@@ -70,13 +69,13 @@ char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uin
         }
 
-		// compute how number of bytes will be linearized
+	// compute how number of bytes will be linearized
```
again, tab

slaperche-scality commented on backend.go:
```diff
@@ -85,7 +84,7 @@ char* decode_fast(int k, char **in, int inlen, char *dest, uint64_t destlen, uin
         }
 
-		// copy in an ordered way all bytes of fragments in the buffer
+	// copy in an ordered way all bytes of fragments in the buffer
```
tab again

slaperche-scality commented on backend.go:
```diff
@@ -5,16 +5,218 @@ package erasurecode
+          if (len_to_copy > dataend - dataoffset) {
+                  len_to_copy = dataend - dataoffset;
+		  }
+		  tot_len_sum += len_to_copy;
```
Those two lines are indented with a mix of tab and spaces.

slaperche-scality commented on backend.go:
```diff
@@ -383,10 +388,12 @@ func (backend *Backend) Encode(data []byte) (*EncodeData, error) {
 		// Note: the 1 << 30 is not really used, the slice is set to a length & a capacity.
-		result[i] = (*[1 << 30]byte)(C.getStrArrayItem(dataFrags, C.int(i)))[:int(fragLength):int(fragLength)]
+		str := cGetArrayItem(dataFrags, i)
+		result[i] = (*[1 << 30]byte)((str))[:int(fragLength):int(fragLength)]
```
I think we should also move this `(*[1 << 30]byte)((str))[:int(fragLength):int(fragLength)]` into an helper.

That would limit the `1 << 30` hack (and its documentation) to a single place, instead of having it at many places (but not always documented).
Also, having a named helper make it more readable (`makeCBackedSlice` is easier to read than this)

slaperche-scality commented on backend.go:
```diff
@@ -492,7 +487,6 @@ func (backend *Backend) DecodeMatrix(frags [][]byte, piecesize int) (*DecodeData
 		// in concurrency without need to lock it access
 		go func(blockNr int) {
 			cFrags := C.makeStrArray(C.int(len(frags)))
-			defer C.freeStrArray(cFrags)
```
Does this have a non-negligible impact on the perf?

- If not, better keep the standard Go approach of using `defer`.
- If yes, would be nice to have the benchmark result before/after (in the commit message) to show the benefits.

slaperche-scality commented on backend_test.go:
```diff
@@ -574,15 +577,15 @@ func BenchmarkDecodeMSlow(b *testing.B) {
+		encoded.Free()
+		decoded.Free()
 	}
+	backend.Close()
```
Maybe use `defer` for those `backend.Close()`?

fferrandis committed:
Fix bad indent

fferrandis committed:
Update GetRangeMatrix

- we need to know fragSize to check that the range won't overflow the
frag len
- update tests and keep only the relevants ones .

fferrandis committed:
update range matrix to deliver offset on linearized data

fferrandis committed:
avoid C allocation and Go pointer aliasing

benoit-a committed:
Add a new bench on reconstructM

Optimize reconstruct (reduce memory allocation)

benoit-a committed:
Improve performance & add more test

benoit-a committed:
Introduce a pool for buffer returned in DecodeMatrix

* add new benchs

benoit-a committed:
add a dedicated pool in the backend

fferrandis commented:
@bert-e approve
